zim/demo-library/demo-generator.py

```diff
@@ -0,0 +1,398 @@
+        (favicons_path / f"{zim_infos['id']}.png").write_bytes(
+            zim_archive.get_illustration_item(  # pyright: ignore [reportUnknownMemberType, reportUnknownArgumentType]
+                48
+            ).content
+        )
+
+    return ZimInfo(path=zim_path, infos=zim_infos)
+
+
+def get_tmp(fpath: Path) -> Path:
+    """path to use to write fpath temporarily"""
+    return Path(
+        tempfile.NamedTemporaryFile(
+            prefix=fpath.stem, suffix=fpath.suffix, delete=False
+        ).name
+    )
+
+
+def swap(tmp: Path, final: Path):
+    """moves* tmp to its final dest via copy + rm to prevent cross-fs errors"""
+    shutil.copy2(tmp, final)
+    tmp.unlink()
+
+
+@contextmanager
+def open_chmod(file: Path, mode: str = "r", chmod: int | None = None):
+    """open a file and then chmod it after use"""
+    with open(file, mode=mode) as fh:
+        yield fh
+    if chmod:
+        file.chmod(chmod)
+
+
+logger.info(f"Downloading demo configuration…")
+resp = requests.get(
+    assets_base_url + "demos.yaml",
+    timeout=REQUESTS_TIMEOUT,
+)
+resp.raise_for_status()
+demo_config.write_bytes(resp.content)
+
+if not demo_dead_html.exists():
+    logger.info(f"Downloading dead.html…")
+    resp = requests.get(
+        assets_base_url + "dead.html",
+        timeout=REQUESTS_TIMEOUT,
+    )
+    resp.raise_for_status()
+    demo_dead_html.write_bytes(resp.content)
+
+if not demo_favicon.exists():
+    logger.info(f"Downloading favicon.ico…")
+    resp = requests.get(
+        assets_base_url + "favicon.ico",
+        timeout=REQUESTS_TIMEOUT,
+    )
+    resp.raise_for_status()
+    demo_favicon.write_bytes(resp.content)
+
+if not demo_dead_kiwix_png.exists():
+    logger.info(f"Downloading dead-kiwix.png…")
+    resp = requests.get(
+        assets_base_url + "dead-kiwix.png",
+        timeout=REQUESTS_TIMEOUT,
+    )
+    resp.raise_for_status()
+    demo_dead_kiwix_png.write_bytes(resp.content)
+
+if not demo_home_tmpl_path.exists():
+    logger.info(f"Downloading demo-home.jinja2…")
+    resp = requests.get(
+        assets_base_url + "demo-home.jinja2",
+        timeout=REQUESTS_TIMEOUT,
+    )
+    resp.raise_for_status()
+    demo_home_tmpl_path.write_bytes(resp.content)
+
+logger.info(f"Preparing…")
+zims: dict[str, ZimInfo] = {}
+
+jinja_env = Environment(
+    autoescape=select_autoescape(), loader=FileSystemLoader(demos_path)
+)
+
+jinja_env.filters["human_size"] = (  # pyright: ignore [reportUnknownMemberType]
+    binary_human_size
+)
+demo_tmpl = jinja_env.get_template(demo_home_tmpl_path.name)
+
+logger.info("Parsing demo configuration…")
+config = Config(**yaml.safe_load(demo_config.read_text()))
+
+# Filter demos based on expired datetime
+now = datetime.date.today()
+config.demos = [demo for demo in config.demos if demo.expired_on > now]
+
+logger.info("Searching for declared ZIMs and extracting infos…")
+for demo in config.demos:
+    for zim_archive in demo.zims:
+        logger.info(f"- {zim_archive}")
```

Comment:
zim_archive represents a path, should be made a `Path`

Reply:
I don't agree, as mentioned in get_ziminfo, it can be an incomplete path

Reply:
Incomplete path is still a path. `Path` doesn't have to exist